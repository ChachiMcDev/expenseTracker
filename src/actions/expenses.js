import { v4 as uuid } from 'uuid'
import database from '../firebase/firebase';

//Expenses actions

//component calls action generator
//action generator returns an object
//component dispatches object to redux
//redux store changes

//*** asych flow     */
//component calls action generator
//action generator returns a function
//component dispatches function ((?) =>   redux does not natively allow functions, redux middlware is used ) 
//////// using redux-thunk as the middleware
//function runs (has the abilit to dispatch other actions and do whaterver it wants)




//ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});


export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = { description, note, amount, createdAt }

        return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};


//REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const startRemoveExpense = ({ id } = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`)
            .remove()
            .then(() => {
                dispatch(removeExpense({ id }))
            });
    };
};





//EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

export const startEditExpense = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
            dispatch(editExpense(id, updates));
        });
    };
};



//SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});


export const startSetExpenses = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses`)
            .once('value')
            .then((snapshot) => {
                const expenses = [];
                snapshot.forEach((childsnapshot) => {
                    expenses.push({
                        id: childsnapshot.key,
                        ...childsnapshot.val()
                    })
                });

                dispatch(setExpenses(expenses));
            });
    };

};


