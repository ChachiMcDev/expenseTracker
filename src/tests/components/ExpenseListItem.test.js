import React from 'react';
import { shallow } from 'enzyme';
import ExpenseListItem from '../../components/ExpenseListItem';
import expenses from '../fixtures/expensesData';



test('expect expense list item to be rendered', ()=>{
    const wrapper = shallow(<ExpenseListItem {...expenses[0]}/> );
    expect(wrapper).toMatchSnapshot();
});