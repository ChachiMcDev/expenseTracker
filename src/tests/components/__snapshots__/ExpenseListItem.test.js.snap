// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expect expense list item to be rendered 1`] = `
<div>
  <[object Object]
    className="list-item"
    to="/edit/0"
  >
    <div>
      <h3
        className="list-item__title"
      >
        yup
      </h3>
      <span
        className="list-item__sub-title"
      >
        Dec 31 1969
      </span>
    </div>
    <h3
      className="list-item__data"
    >
      $1.95
       
    </h3>
  </[object Object]>
</div>
`;
